dm/worker/task_checker.go

```diff
@@ -0,0 +1,315 @@
+		return
+	}
+
+	if tsc.cancel != nil {
+		tsc.cancel()
+	}
+	tsc.wg.Wait()
+}
+
+func (tsc *realTaskStatusChecker) run() {
+	tsc.ctx, tsc.cancel = context.WithCancel(context.Background())
+	tsc.closed.Set(closedFalse)
+	ticker := time.NewTicker(tsc.cfg.CheckInterval)
+	defer ticker.Stop()
+	for {
+		select {
+		case <-tsc.ctx.Done():
+			tsc.l.Info("worker task checker exits")
+			return
+		case <-ticker.C:
+			tsc.check()
+		}
+	}
+}
+
+// isResumableError checks the error message and returns whether we need to
+// resume the task and retry
+func isResumableError(err *pb.ProcessError) bool {
+	// not elegant code, because TiDB doesn't expose some error
+	unsupportedDDLMsgs := []string{
+		"can't drop column with index",
+		"unsupported add column",
+		"unsupported modify column",
+		"unsupported modify",
+		"unsupported drop integer primary key",
+	}
+	if err.Type == pb.ErrorType_ExecSQL {
+		for _, msg := range unsupportedDDLMsgs {
+			if strings.Contains(err.Msg, msg) {
+				return false
+			}
+		}
+	}
+	return true
+}
+
+func (tsc *realTaskStatusChecker) getResumeStrategy(stStatus *pb.SubTaskStatus, duration time.Duration) ResumeStrategy {
+	// task that is not paused or paused manually, just ignore it
+	if stStatus == nil || stStatus.Stage != pb.Stage_Paused || stStatus.Result == nil || stStatus.Result.IsCanceled {
+		return ResumeIgnore
+	}
+
+	// TODO: use different strategies based on the error detail
+	for _, processErr := range stStatus.Result.Errors {
+		if !isResumableError(processErr) {
+			return ResumeNoSense
+		}
+	}
+
+	// auto resume interval does not exceed backoff duration, skip this paused task
+	if time.Since(tsc.bc.latestResumeTime[stStatus.Name]) < duration {
+		return ResumeSkip
+	}
+
+	return ResumeDispatch
+}
+
+func (tsc *realTaskStatusChecker) check() {
+	allSubTaskStatus := tsc.w.getAllSubTaskStatus()
+
+	defer func() {
+		// cleanup outdated tasks
+		for taskName := range tsc.bc.backoffs {
+			_, ok := allSubTaskStatus[taskName]
+			if !ok {
+				tsc.l.Debug("remove task from checker", zap.String("task", taskName))
+				delete(tsc.bc.backoffs, taskName)
+				delete(tsc.bc.latestPausedTime, taskName)
+				delete(tsc.bc.latestBlockTime, taskName)
+				delete(tsc.bc.latestResumeTime, taskName)
+			}
+		}
+	}()
+
+	for taskName, stStatus := range allSubTaskStatus {
+		bf, ok := tsc.bc.backoffs[taskName]
+		if !ok {
+			bf, _ = backoff.NewBackoff(tsc.cfg.BackoffFactor, tsc.cfg.BackoffJitter, tsc.cfg.BackoffMin, tsc.cfg.BackoffMax)
+			tsc.bc.backoffs[taskName] = bf
+			tsc.bc.latestPausedTime[taskName] = time.Now()
+			tsc.bc.latestResumeTime[taskName] = time.Now()
+		}
+		duration := bf.Current()
+		strategy := tsc.getResumeStrategy(stStatus, duration)
+		switch strategy {
+		case ResumeIgnore:
+			if time.Since(tsc.bc.latestPausedTime[taskName]) > tsc.cfg.BackoffRollback {
+				bf.Rollback()
+				// after each rollback, reset this timer
+				tsc.bc.latestPausedTime[taskName] = time.Now()
```

Comment:
misleading name and trick usage, but I have no better solution. :joy: (no need to address